tooltester/src/main/java/io/dockstore/tooltester/S3Client.java

```diff
@@ -0,0 +1,116 @@
+package io.dockstore.tooltester;
+
+import java.io.ByteArrayInputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import javax.ws.rs.core.MediaType;
+
+import com.amazonaws.client.builder.AwsClientBuilder;
+import com.amazonaws.regions.Regions;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.AmazonS3ClientBuilder;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.PutObjectRequest;
+
+/**
+ * @author gluu
+ * @since 18/04/19
+ */
+public class S3Client {
+    private static final int MAX_TOOL_ID_STRING_SEGMENTS = 5;
+    private static final int TOOL_ID_REPOSITORY_INDEX = 3;
+    private static final int TOOL_ID_TOOLNAME_INDEX = 4;
+    private String bucketName;
+    private AmazonS3 s3;
+
+    public S3Client() {
+        TooltesterConfig tooltesterConfig = new TooltesterConfig();
+        String s3Endpoint = tooltesterConfig.getS3Endpoint();
+        bucketName = tooltesterConfig.getS3Bucket();
+        AwsClientBuilder.EndpointConfiguration endpointConfiguration = new AwsClientBuilder.EndpointConfiguration(s3Endpoint,
+                Regions.US_EAST_1.getName());
+        s3 = AmazonS3ClientBuilder.standard().withEndpointConfiguration(endpointConfiguration).build();
+    }
+
+    /**
+     * This essentially generates the filePath of the log that will be stored on S3
+     *
+     * @param toolId       The GA4GH Tool ID
+     * @param versionName  The GA4GH ToolVersion name
+     * @param testFilePath The file that was tested (Dockerfile, test.json, etc)
+     * @param runner       The runner used to test (cwltool, cromwell, etc)
+     * @param startTime    The start time in milliseconds since epoch
+     * @return S3 key (file path)
+     * @throws UnsupportedEncodingException Could not endpoint string
+     */
+    private static String generateKey(String toolId, String versionName, String testFilePath, String runner, String startTime)
+            throws UnsupportedEncodingException {
+        List<String> pathList = new ArrayList<>();
+        pathList.add(convertToolIdToPartialKey(toolId));
+        pathList.add(URLEncoder.encode(versionName, StandardCharsets.UTF_8.toString()));
+        pathList.add(
+                URLEncoder.encode(testFilePath.replaceFirst("^Build ", "").replaceFirst("^Test ", ""), StandardCharsets.UTF_8.toString()));
```

Comment:
A comment on why testFilePath starts with `Build`, `Test`, or `BuildTest` and why you're doing this could be useful.